### PR TITLE
CI: Test on newer Pythons; enable Python dev mode

### PR DESCRIPTION
**Pull request by akx**
See https://docs.python.org/3.12/library/devmode.html#devmode for what `PYTHONDEVMODE` does.

### CODE REVIEW

**Comment by akx** [comment]
@PierreF Rebased, ready for review.

File: .github/workflows/tox.yml [hunk 1]
```diff
@@ -6,17 +6,27 @@ jobs:
...
       - if: matrix.python == '3.8'
         run: TOXENV=py38 tox
       - if: matrix.python == '3.9'
         run: TOXENV=py39 tox
+      - if: matrix.python == '3.10'
+        run: TOXENV=py310 tox
+      - if: matrix.python == '3.11'
+        run: TOXENV=py311 tox
+      - if: matrix.python == '3.12'
+        run: TOXENV=py312 tox
```
Comment (cclauss) [inline .github/workflows/tox.yml:32]: Let's use the `- run: tox -e py` as discussed in the GHA docs and #777

File: tox.ini [hunk 2]
```diff
@@ -11,8 +11,10 @@ allowlist_externals =
     make
 commands =
     ruff .
-    python setup.py test
+    pytest
     make -C test test
+env =
+    PYTHONDEVMODE=1
```
Comment (cclauss) [inline tox.ini:17]: How does this work?

**Review by cclauss — APPROVED** [review]
Love the addition of `PYTHONDEVMODE` -- Cool!

**Comment by PierreF** [comment]
The PR looks fine, we just broke "python setup.py test" command.
I'm not sure whether is really used (I think the the release process, which might use use). Anyway, I'll have a look at #774 which should both cause few conflict with this PR and remove "setup.py" completely.

**Comment by cclauss** [comment]
https://setuptools.pypa.io/en/latest/deprecated/commands.html
> running `python setup.py` directly as a script is considered __deprecated__. This also means that users should avoid running commands directly via `python setup.py <command>`.

**Comment by akx** [comment]
Yeah, breaking `setup.py test` was on purpose – there's no reference to it anywhere in the repo, so I figured it wasn't really needed (anymore)...

**Comment by PierreF** [comment]
Confirmed that setup.py is not used for release, at least not how I do release (which I should document somewhere).

My process is currently:
* testing by calling tox (inside a Docker image python:3.x)
* testing linter by calling tox
* tag & push tag
* build and upload to PyPi (using make upload)

Therefor this PR seems fine for me.

**Comment by akx** [comment]
> My process is currently: ...

If it helps, the procedure I've built for python-babel is pretty slick these days – everything is done on GitHub Actions, and I just okay the release (via the `release` GHA environment requiring an accept gate).

**Comment by PierreF** [comment]
That true that making a release on CI is better than locally, as it ensure your local environment don't affect the release. I take not of your workflow and will consider it.

**Pull request merged**